File: 03--gear-ratios/a.py
#!/usr/bin/env python3
'''
Usage:
    ./a.py PATH_TO_INPUT_FILE
    ./b.py PATH_TO_INPUT_FILE
'''

import sys
import re
import string
from collections import namedtuple


PartNumber = namedtuple('PartNumber', 'value left_endpoint right_endpoint')


def main():
    schematic = open(sys.argv[1]).read().splitlines()
    part_numbers = []
    for r, row in enumerate(schematic):
        for start, end in find_numbers(row):
            n = int(row[start : end+1])
            part_numbers.append(
                PartNumber(n, (r, start), (r, end))
            )

    answer = sum(
        each.value
        for each in part_numbers
        if any(is_symbol(neighbor) for neighbor in neighbors(schematic, each))
    )

    print(answer)


def is_symbol(ch):
    return (
        ch not in string.digits
        and ch != '.'
    )


def find_numbers(line):
    start = None
    results = []
    for i, ch in enumerate(line + '!'):
        if ch not in string.digits and start is not None:
            results.append((start, i - 1))
            start = None
        elif ch in string.digits and start is None:
            start = i
    return results


def neighbors(schematic, part_number):
    HEIGHT = len(schematic)
    WIDTH = len(schematic[0])

    positions = []
    R, start = part_number.left_endpoint
    _, end = part_number.right_endpoint
    for c in range(start - 1, end + 2):
        positions.append((R - 1, c))
        positions.append((R + 1, c))
    positions.append((R, start - 1))
    positions.append((R, end + 1))

    for r, c in positions:
        if (
            0 <= r < HEIGHT and
            0 <= c < WIDTH
        ):
            yield schematic[r][c]


if __name__ == '__main__':
    main()

File: 03--gear-ratios/a.hs
#!/usr/bin/env runhaskell
-- Usage:
--   ./a.hs PATH_TO_INPUT_FILE
--   ./b.hs PATH_TO_INPUT_FILE

import Data.Char (isDigit)
import Data.Function ((&), on)
import Data.List (groupBy, elem, nub, sort)
import System.Environment (getArgs)

type Position = (Int, Int)
type NumberPosition = ((Int, Int), (Int, Int))

main = do
  [inputPath] <- getArgs
  text <- readFile inputPath
  print $ solve text

solve text =
  numberPositions
  & filter (\pos -> any isSymbol $ neighbors pos)
  & map positionsInNumber
  & map (\positions -> map get positions)
  & map read
  & sum
  where
    myLines = lines text
    get (r, c) = myLines !! r !! c
    numberPositions = findNumbers myLines
    height = length myLines
    width = length $ head myLines
    dimensions = (height, width)
    neighbors numberPosition =
      neighborPositions dimensions numberPosition
      & map get

positionsInNumber numberPosition@((r1,c1),(r2,c2)) =
  [(r1, c) | c <- [c1..c2]]

isSymbol ch = (not $ isDigit ch) && (ch /= '.')

neighborPositions dimensions@(height, width) numberPosition =
  myPositionsInNumber
  & concatMap
      (\(r, c) ->
        [
          (r - 1, c - 1), (r - 1, c), (r - 1, c + 1),
          (r,     c - 1), (r,     c), (r,     c + 1),
          (r + 1, c - 1), (r + 1, c), (r + 1, c + 1)
        ]
      )
  & sort
  & nub
  & filter (\pos -> not $ pos `elem` myPositionsInNumber)
  & filter inBounds
  where
    myPositionsInNumber = positionsInNumber numberPosition
    inBounds (r, c) =
      0 <= r && r < height &&
      0 <= c && c < width

findNumbers myLines =
  zip
    [0..]
    (map findNumbersInLine myLines)
  & concatMap
      (\(r, groups) ->
        map
          (\(groupStartCol, groupEndCol) ->
            ((r, groupStartCol), (r, groupEndCol)) :: NumberPosition
          )
          groups
      )

findNumbersInLine line =
  line
  & zip [0..]
  & groupBy ((==) `on` (isDigit . snd))
  & filter (\group -> isDigit (snd (head group)))
  & map (\group -> map fst group)
  & map (\group -> (head group, last group))
